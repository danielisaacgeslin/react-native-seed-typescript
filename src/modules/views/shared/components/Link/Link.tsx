import React from 'react';
import { TouchableOpacity, Text, Linking } from 'react-native';

import { Logger } from '../../../../services/Logger';
import Styles from './styles';

export interface ILinkProps {
  url: string;
  text?: string;
}

export default class Link extends React.PureComponent<ILinkProps> {
  public static linking: Linking = Linking;

  public render() {
    const { url, text } = this.props;
    return (
      <TouchableOpacity onPress={this.onPress}>
        <Text style={Styles.link}>{text || url}</Text>
      </TouchableOpacity>
    );
  }

  private onPress = async () => {
    const { url } = this.props;
    try {
      const supported = await Link.linking.canOpenURL(url);
      if (supported) Link.linking.openURL(url);
    } catch (e) {
      Logger.error('Error linking', e);
    }
  };
}
