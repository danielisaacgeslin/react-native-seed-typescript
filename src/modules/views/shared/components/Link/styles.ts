import { StyleSheet } from 'react-native';

import { STYLE } from '../../../../../constants';

export default StyleSheet.create({
  link: {
    textDecorationLine: 'underline',
    letterSpacing: 0.25,
    color: STYLE.COLOR.SECONDARY,
    fontSize: STYLE.FONT.SIZE.PARAGRAPH,
    fontWeight: STYLE.FONT.WEIGHTS.MEDIUM,
    lineHeight: STYLE.FONT.SIZE.PARAGRAPH
  }
});
