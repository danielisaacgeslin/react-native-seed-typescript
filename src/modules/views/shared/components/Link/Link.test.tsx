import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Link from './Link';

describe('Link', () => {
  let props;
  beforeEach(() => {
    props = {
      url: 'https://url.jpg'
    };
  });
  it('should render', () => {
    const rendered = renderer.create(<Link {...props} />).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('should render with text', () => {
    props = {
      url: 'https://url.jpg',
      text: 'This is a Link'
    };
    const rendered = renderer.create(<Link {...props} />).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('should open a link', async () => {
    const url = 'https://google.com';
    const text = 'Google';
    Link.linking = {
      canOpenURL: jest.fn().mockReturnValue(Promise.resolve(true)),
      openURL: jest.fn()
    } as any;
    const wrapper = shallow(<Link {...props} />);
    wrapper.setProps({ url, text });
    await (wrapper
      .find('TouchableOpacity')
      .first()
      .prop('onPress') as any)();
    expect(Link.linking.canOpenURL).toBeCalledWith(url);
    expect(Link.linking.openURL).toBeCalledWith(url);
  });

  it('should not open a link', async () => {
    const url = 'https://google.com';
    const text = 'Google';
    Link.linking = {
      canOpenURL: jest.fn().mockReturnValue(Promise.resolve(false)),
      openURL: jest.fn()
    } as any;
    const wrapper = shallow(<Link {...props} />);
    wrapper.setProps({ url, text });
    await (wrapper
      .find('TouchableOpacity')
      .first()
      .prop('onPress') as any)();
    expect(Link.linking.canOpenURL).toBeCalledWith(url);
    expect(Link.linking.openURL).not.toBeCalledWith(url);
  });

  it('should catch error', () => {
    const url = 'https://google.com';
    Link.linking = {
      canOpenURL: () => {
        throw new Error('Error');
      }
    } as any;
    const wrapper = shallow(<Link {...props} />);
    wrapper.setProps({ url });
    (wrapper
      .find('TouchableOpacity')
      .first()
      .prop('onPress') as any)();
  });
});
