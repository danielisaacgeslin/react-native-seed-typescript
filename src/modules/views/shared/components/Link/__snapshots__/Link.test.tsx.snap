// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should render 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="link"
>
  <Text
    style={
      Object {
        "color": "#5BC9EB",
        "fontSize": 15,
        "fontWeight": "600",
        "letterSpacing": 0.25,
        "lineHeight": 15,
        "textDecorationLine": "underline",
      }
    }
  >
    https://url.jpg
  </Text>
</View>
`;

exports[`Link should render with text 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="link"
>
  <Text
    style={
      Object {
        "color": "#5BC9EB",
        "fontSize": 15,
        "fontWeight": "600",
        "letterSpacing": 0.25,
        "lineHeight": 15,
        "textDecorationLine": "underline",
      }
    }
  >
    This is a Link
  </Text>
</View>
`;
